zip-proportional-output-fee-mechanism-pofm.rst

```diff
@@ -0,0 +1,295 @@
+This has resulted in high output transactions with 1,100 outputs costing the same as 
+transactions with 2 outputs.
+
+
+Requirements for gathering consensus
+------------------------------------
+
+Wallet developers SHOULD update the fees to the proposed formula under Specification section
+by Madars Virza and Kris Nuttycombe [#madars-1]_ 
+
+Where #inputs and #outputs also take into account transparent inputs and outputs. 
+Otherwise, the fee structure (if not otherwise changed) will preferentially encourage 
+usage of the transparent part of the protocol.
+
+The change to the conventional transaction fees must be undertaken soon
+as the Zcash network has been under heavy load with high-output transactions while 
+regular shielded transactions with 2 outputs are affected when relying on current 
+light wallet infrastructure.
+
+The following parties need to be part of the consensus:
+
+* The technical aspects of a changing conventional fee based on outputs 
+need to be evaluated.
+* A guarantee from mining groups is required to include the updated POFM 
+transactions in the next block.
+* Wallet developers need to update the software to use the new fee.
+* Zcash documentation and community outreach must be undertaken to
+make the change known.
+
+
+Requirements for adoption
+-------------------------
+
+The change to the conventional transaction fees should be undertaken soon
+as it gets difficult to gain consensus with the growth in the network
+of wallets, exchanges, miners, and third parties involved.
+
+The following parties need to be part of the consensus:
+
+* Support from mining groups is required to include the updated conventional
+  fee transactions in the next block.
+* Wallet developers need to provide a commitment to update the software to use
+  the new fee.
+* Zcash documentation and community outreach must be undertaken to make the
+  change known.
+* Changes to librustzcash crates to:
+  * expose minimum recommended fee for a given number of recipients
+  * implement default fee equation and corresponding unit tests
+  * integration of such mechanism to Transaction Builder APIs
+* Changes to the Zebra mempool's conventional fee calculation
+* adopt `librustzcash` changes to mobile SDK repos:
+  * ZcashLightClientKit
+  * zcash-light-client-ffi
+  * zcash-android-wallet-sdk
+
+
+Specification
+=============
+
+This specification defines three parameters that are used to calculate the
+conventional fee:
+
+=================== ============================
+Parameter           Units
+=================== ============================
+`base_fee`          zatoshis
+`marginal_fee`      zatoshis per input or output
+`grace_window_size` inputs or outputs
+=================== ============================
+
+Wallets implementing this specification SHOULD use a conventional fee
+calculated in zatoshis per the following formula::
+
+    base_fee + marginal_fee * max(0, inputs + outputs - grace_window_size)
```

Comment:
We’re considering replacing `inputs + outputs` with a generalization of “actions”, roughly `max(inputs, outputs)` here are the pros and cons we’re currently aware of:

* Pro: It doesn't disincentivize the use of Orchard. With `+`, a Sapling->Sapling tx (with change) would have an arity of 3, whereas the same Orchard->Orchard tx would have an arity of 4. With `max`, both would have an arity of 2 (and this gap gets larger as the arity increases – almost all high-arity transactions are many→few or few→many).
* Con: It could motivate miners to prioritize non-Orchard transactions, as the Sapling tx would effectively pay for 4 “slots” while only using three and the Orchard tx would use all four, so it’s less work to process Sapling for the same fee. However, fees and proving/verifying are minuscule compared to rewards and mining (for now), so this effect seems neglible.
* Pro: Notes with value `<= marginal_fee` would be spendable via the “free” input slots when tx has more outputs than inputs. This would be true in Orchard regardless of the fee mechanism, but requires something like `max` in order to work in other pools. https://github.com/zcash/zips/pull/631#issuecomment-1252599792 mentions this as a benefit of a non-zero `grace_window`, but that only works in tx with low arity. This would be more general.

We also want to consider how cross-pool tx are affected. E.g., a Sapling→Orchard tx (1 input note, with Sapling change) would have an arity of 4 under `+` or an arity of 2 under `max`, which is the same as the equivalent all-Orchard tx, but this becomes 5 or 3, respectively, if the tx requires spending two Sapling notes to cover the cost, which is more than the fee for an equivalent tx within any single pool.